data/src/main/java/com/linkedin/data/schema/validation/ValidateDataAgainstSchema.java

```diff
@@ -278,13 +292,82 @@ protected void recurseRecord(DataElement element, RecordDataSchema schema, DataM
       {
         String key = entry.getKey();
         RecordDataSchema.Field field = schema.getField(key);
+        Object value = entry.getValue();
         if (field != null)
         {
-          Object value = entry.getValue();
           DataSchema childSchema = field.getType();
           childElement.setValueNameSchema(value, key, childSchema);
           validate(childElement, childSchema, value);
         }
+        else
+        {
+          childElement.setValueNameSchema(value, key, null);
+          handleUnrecognizedField(childElement);
+        }
+      }
+    }
+
+    /**
+     * Holds a the element of a {@link com.linkedin.data.DataMap} field for later removal.
+     *
+     * This is a specialized version of {@link com.linkedin.data.it.Remover.ToRemove} which is and
+     * should remain private, so could not be used here.
+     */
+    private static class FieldToTrim
+    {
+      private FieldToTrim(DataMap parent, String fieldName)
+      {
+        assert(!parent.isReadOnly());
+        _parent = parent;
+        _fieldName = fieldName;
+      }
+
+      private void trim()
+      {
+        if (_parent.isReadOnly())
+        {
+          throw new ConcurrentModificationException("Map marked as read-only during validation.");
```

Comment:
This should be unsupported operation exception. This is what Java uses for such scenarios (i.e unmodifiable list/map). What do you think?